docs/features/async-streams.md

```diff
@@ -91,141 +90,41 @@ The state machine for an async-iterator method primarily implements `IAsyncEnume
-- to support handling a `yield return` statement, which saves the current value and fulfill the promise (if active),
-- to support handling a `yield break` statement, which resets the promise (if active) and fulfills it with result `false`.
+- to support handling a `yield return` statement, which saves the current value and fulfills the promise with result `true`,
+- to support handling a `yield break` statement, which fulfills the promise with result `false`,
+- to exit the method, which fulfills the promise with result `false`,
+- to the handling of exceptions, to set the exception into the promise.
+(The handling of an `await` is unchanged)
 
 This is reflected in the implementation, which extends the lowering machinery for async methods to:
-1. properly signal through the promise (update in various methods in `AsyncMethodToStateMachineRewriter`, such as `VisitAwaitExpression`)
-2. handle `yield return` and `yield break` statements (add methods `VisitYieldReturnStatement` and `VisitYieldBreakStatement` to `AsyncMethodToStateMachineRewriter`),
-3. produce additional state and logic for the promise itself (we use `AsyncIteratorRewriter` instead of `AsyncRewriter` to drive the lowering, and produces the other members: `WaitForNextAsync`, `TryGetNext`, `DisposeAsync`, and some members supporting the resettable `ValueTask`, namely `GetResult`, `SetStatus`, `OnCompleted` and `Value.get`).
-
-The contract of the `MoveNext()` method is that it returns either:
-- in completed state,
-- leaving the promise inactive (when started with an inactive promise and a value is immediately available),
-- with an exception (when started with an inactive promise and an exception is thrown),
-- an active promise, which will later be fulfilled (with `true`, `false` or an exception).
-
-If the promise is active:
-- the builder is running the `MoveNext()` logic,
-- a call to `WaitForNextAsync` will not move the state machine forward (ie. it won't call `MoveNext()`),
-- a call to `TryGetNext` APIs will throw.
-
-**open issue**: The compiler leverages existing BCL types (including some recently added types from the `System.Threading.Tasks.Extensions` NuGet package) in the state machine it generates. But as part of this feature, we may introduce some additional BCL types, so that the state machine can be further simplified and optimized.
+1. handle `yield return` and `yield break` statements (add methods `VisitYieldReturnStatement` and `VisitYieldBreakStatement` to `AsyncMethodToStateMachineRewriter`),
+2. produce additional state and logic for the promise itself (we use `AsyncIteratorRewriter` instead of `AsyncRewriter` to drive the lowering, and produces the other members: `MoveNextAsync`, `Current`, `DisposeAsync`, and some members supporting the resettable `ValueTask`, namely `GetResult`, `SetStatus`, `OnCompleted` and `Value.get`).
 
 ```C#
-ValueTask<bool> WaitForNextAsync()
+ValueTask<bool> MoveNextAsync()
 {
```

Comment:
:memo: I have a follow-up issue to guard against calling `MoveNextAsync()` when the promise is pending. [[https://github.com/dotnet/roslyn/issues/30109](https://github.com/dotnet/roslyn/issues/30109)](https://github.com/dotnet/roslyn/issues/30109)